File: golang/linear/stack_queue.go
package linear

type Stack[T any] struct {
	data []T
}

func (s *Stack[T]) Push(v T) {
	s.data = append(s.data, v)
}

func (s *Stack[T]) Pop() (T, bool) {
	var zero T
	if len(s.data) == 0 {
		return zero, false
	}
	v := s.data[len(s.data)-1]
	s.data = s.data[:len(s.data)-1]
	return v, true
}

func (s *Stack[T]) Peek() (T, bool) {
	var zero T
	if len(s.data) == 0 {
		return zero, false
	}
	return s.data[len(s.data)-1], true
}

func (s *Stack[T]) Size() int { return len(s.data) }

func (s *Stack[T]) IsEmpty() bool { return len(s.data) == 0 }

type Queue[T any] struct {
	data []T
}

func (q *Queue[T]) Offer(v T) {
	q.data = append(q.data, v)
}

func (q *Queue[T]) Poll() (T, bool) {
	var zero T
	if len(q.data) == 0 {
		return zero, false
	}
	v := q.data[0]
	q.data = q.data[1:]
	return v, true
}

func (q *Queue[T]) Peek() (T, bool) {
	var zero T
	if len(q.data) == 0 {
		return zero, false
	}
	return q.data[0], true
}

func (q *Queue[T]) Size() int { return len(q.data) }

func (q *Queue[T]) IsEmpty() bool { return len(q.data) == 0 }

File: python/stack_queue.py
from collections import deque


class Stack:
    def __init__(self):
        self._data = []

    def push(self, value):
        self._data.append(value)

    def pop(self):
        if self.is_empty():
            raise IndexError("stack is empty")
        return self._data.pop()

    def peek(self):
        if self.is_empty():
            raise IndexError("stack is empty")
        return self._data[-1]

    def size(self):
        return len(self._data)

    def is_empty(self):
        return len(self._data) == 0


class Queue:
    def __init__(self):
        self._data = deque()

    def offer(self, value):
        self._data.append(value)

    def poll(self):
        if self.is_empty():
            raise IndexError("queue is empty")
        return self._data.popleft()

    def peek(self):
        if self.is_empty():
            raise IndexError("queue is empty")
        return self._data[0]

    def size(self):
        return len(self._data)

    def is_empty(self):
        return len(self._data) == 0
